bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/discovery/HeosPlayerDiscovery.java

```diff
@@ -133,52 +145,123 @@ protected void startScan() {
+     * @return a HashMap with all HEOS Player in the network
+     */
+    private @Nullable Map<Integer, Player> getNewPlayers() {
+        try {
+            playerMapNew.clear();
+
+            for (Player player : bridge.getPlayers()) {
+                playerMapNew.put(player.playerId, player);
+                removedPlayerMap = comparePlayerMaps(playerMapNew, playerMapOld);
+                playerMapOld.clear();
+                playerMapOld.putAll(playerMapNew);
+            }
+            return playerMapNew;
+        } catch (IOException | Telnet.ReadException e) {
+            logger.debug("Failed getting/processing groups", e);
+            return null;
+        }
+    }
+
+    /**
+     * This method searches for all groups which are on the HEOS network
+     * and returns a {@code Map<String gid, HeosGroup heosGroup>}.
+     * Before calling this method a connection via a bridge has to be
+     * established
+     *
+     * @return a HashMap with all HEOS groups
+     * @throws Telnet.ReadException
+     * @throws IOException
+     */
+    private @Nullable Map<String, Group> getNewGroups() {
+        groupMapNew.clear();
+        removedGroupMap.clear();
+
+        try {
+            Group[] groups = bridge.getGroups();
+
+            if (groups.length == 0) {
+                removedGroupMap = compareGroupMaps(groupMapNew, groupMapOld);
+                groupMapOld.clear();
+                return groupMapNew;
+            }
+
+            for (Group group : groups) {
+                logger.debug("Found: Group {} with {} Players", group.name, group.players.size());
+                groupMapNew.put(HeosGroup.calculateGroupMemberHash(group), group);
+                removedGroupMap = compareGroupMaps(groupMapNew, groupMapOld);
+                groupMapOld.clear(); // clear the old map so that only the currently available groups are added in the
+                                     // next
+                // step.
+                groupMapOld.putAll(groupMapNew);
+            }
+            return groupMapNew;
+        } catch (IOException | Telnet.ReadException e) {
+            logger.debug("Failed getting/processing groups", e);
+            return null;
+        }
+    }
+
+    private Map<String, Group> compareGroupMaps(Map<String, Group> mapNew, Map<String, Group> mapOld) {
+        Map<String, Group> removedItems = new HashMap<>();
+        for (String key : mapOld.keySet()) {
+            if (!mapNew.containsKey(key)) {
+                removedItems.put(key, mapOld.get(key));
+            }
+        }
+        return removedItems;
+    }
+
+    private Map<Integer, Player> comparePlayerMaps(Map<Integer, Player> mapNew, Map<Integer, Player> mapOld) {
+        Map<Integer, Player> removedItems = new HashMap<>();
+        for (Integer key : mapOld.keySet()) {
+            if (!mapNew.containsKey(key)) {
+                removedItems.put(key, mapOld.get(key));
+            }
+        }
+        return removedItems;
+    }
```

Comment:
These two methods can be merged using generics.
```suggestion
     private <K,V> Map<K, V> compareMaps(Map<K, V> mapNew, Map<K, V> mapOld) {
        Map<K, V> removedItems = new HashMap<>();
        for (Integer key : mapOld.keySet()) {
            if (!mapNew.containsKey(key)) {
                removedItems.put(key, mapOld.get(key));
            }
        }
        return removedItems;
    }
```

Reply:
Will do, did not realize you already wrote it :-)